### PR TITLE
Fix casting logic for 0d CPU tensors in CUDA ops

### LINKED ISSUE #11795: [TensorIterator] bug when performing inter-scalar ops on the GPU
When performing operations between scalar tensors of different types, type promotion happens in TensorIterator. But it seems that the results of inter-type operations are not correct when performing operations on the GPU.

The following code gives unexpected results on the GPU. CPU works fine
```python
import torch

torch.__version__
# '1.0.0a0+91b6458', master from today

a = torch.tensor(1.5, device='cuda')  # float tensor
b = torch.tensor(1, device='cuda')  # int64 tensor
print(a * b)
# tensor(2.8026e-45, device='cuda:0')

a = torch.tensor(1.5, device='cuda')  # float tensor
b = torch.tensor(True, device='cuda') # uint8 tensor
print(a * b)
# tensor(2.8026e-45, device='cuda:0')
print(a + b)
# tensor(1.5000, device='cuda:0')
print(a - b)
# tensor(1.5000, device='cuda:0')
```

cc @colesbury

### PR DESCRIPTION
**Pull request by colesbury**
Previously, we didn't cast any 0-dim tensors used in CUDA operations. We
can only avoid the casts for 0-dim CPU tensors used in CUDA operations.

Fixes #11795

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.cpp [hunk 1]
```diff
@@ -95,15 +109,11 @@ void TensorIterator::compute_common_type() {
...
-        if (op.tensor->dim() == 0) {
-          if (type.backend() != at::Backend::CUDA) {
-            cast_tensors_.emplace_back(op.tensor->toType(type));
-            op.tensor = &(cast_tensors_.back());
-          }
-        } else {
-          op.needs_cast = true;
-        }
+      op.needs_cast = needs_cast(*op.tensor, type);
+      if (op.needs_cast && op.tensor->dim() == 0) {
```
Comment (driazati) [inline aten/src/ATen/native/TensorIterator.cpp]: What's the reason it only casts for 0 dim tensors and not for any number of dimensions?

Comment (colesbury) [reply aten/src/ATen/native/TensorIterator.cpp]: Eventually, we'll allow casts for any number of dimensions probably using the logic described here:
https://github.com/pytorch/pytorch/issues/9515

For now, we only support casts for zero-dim tensors because it's the minimum necessary to preserve backwards compatibility.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -69,6 +69,20 @@ compute_result_type(at::ArrayRef<OperandInfo> operands, const F& predicate) {
   return std::make_tuple(result_type, backend);
 }
 
+static bool needs_cast(const Tensor& tensor, const Type& dst_type) {
+  if (!tensor.defined() || dst_type == tensor.type()) {
+    return false;
+  }
+  if (dst_type.backend() == Backend::CUDA &&
```
Comment (gchanan) [inline aten/src/ATen/native/TensorIterator.cpp]: do you actually mean for these to be backends or do you mean for them to be device types?  Types have `device_type` now.

File: test/test_cuda.py [hunk 3]
```diff
@@ -903,6 +908,25 @@ def test_type_conversions(self):
...
+        self.assertEqual(x * y, 4.5)
+        self.assertEqual(y * x, 4.5)
+        with self.assertRaisesRegex(RuntimeError, 'expected type'):
+            y *= x
+        x *= y
+        self.assertEqual(x, 4.5)
+
+        x = torch.tensor(1.5, device='cuda', dtype=torch.float16)
+        self.assertEqual(x * y, 4.5)
+        # half * int currently promotes to double
```
Comment (gchanan) [inline test/test_cuda.py]: change this?

**Review by gchanan — APPROVED** [review]

File: test/test_cuda.py (hunk 3 above)
Comment (colesbury) [reply test/test_cuda.py]: I'll make that change in a subsequent PR
